llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -21553,6 +21553,56 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
       JointMemOpVT = EVT::getIntegerVT(Context, SizeInBits);
     }
 
+    auto HasCallInLdStChain = [](SmallVectorImpl<MemOpLink> &StoreNodes,
+                                 SmallVectorImpl<MemOpLink> &LoadNodes,
+                                 unsigned NumStores) {
+      for (unsigned i = 0; i < NumStores; ++i) {
+        StoreSDNode *St = cast<StoreSDNode>(StoreNodes[i].MemNode);
+        SDValue Val = peekThroughBitcasts(St->getValue());
+        LoadSDNode *Ld = cast<LoadSDNode>(Val);
+        assert(Ld == LoadNodes[i].MemNode && "Load and store mismatch");
```

Comment:
Nit, can we remove the peekThroughBitcasts in release builds if we swap the assertion around? Something like

```suggestion
        StoreSDNode *St = cast<StoreSDNode>(StoreNodes[i].MemNode);
        StoreSDNode *Ld = cast<LoadSDNode>(LoadNodes[i].MemNode);
        assert(Ld == peekThroughBitcasts(St->getValue()) && "Load and store mismatch");
```

Reply:
how about the following?
```
-        SDValue Val = peekThroughBitcasts(St->getValue());
-        LoadSDNode *Ld = cast<LoadSDNode>(Val);
-        assert(Ld == LoadNodes[i].MemNode && "Load and store mismatch");
+        LoadSDNode *Ld = cast<LoadSDNode>(LoadNodes[i].MemNode);
+        assert(Ld == cast<LoadSDNode>(peekThroughBitcasts(St->getValue())) &&
+               "Load and store mismatch");
```

Reply:
Argh woops, yes Ld should be LoadSDNode not StoreSDNode!